src/processor/call_handler.rs

```diff
@@ -0,0 +1,130 @@
+/// This instruction is meant to be called via CPI with the owning program signing for the
+/// delegated account.
+pub fn process_call_handler(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: &[u8],
+) -> ProgramResult {
+    let args = CallHandlerArgs::try_from_slice(data)?;
+
+    let accounts_iter = &mut accounts.iter();
+
+    let validator = next_account_info(accounts_iter)?;
+    let validator_fees_vault = next_account_info(accounts_iter)?;
+    let destination_program = next_account_info(accounts_iter)?;
+    let escrow_authority_account = next_account_info(accounts_iter)?;
+    let escrow_account = next_account_info(accounts_iter)?;
+
+    // verify account is a signer
+    load_signer(validator, "validator")?;
+    // verify signer is a registered validator
+    load_initialized_validator_fees_vault(validator, validator_fees_vault, true)?;
+    // Check if destination program is executable
+    if !destination_program.executable {
+        msg!(
+            "{} program is not executable: destination program",
+            destination_program.key
+        );
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    // verify passed escrow_account derived from delegated_account
+    let escrow_seeds: &[&[u8]] =
+        ephemeral_balance_seeds_from_payer!(escrow_authority_account.key, args.escrow_index);
+    let escrow_bump = load_pda(
+        escrow_account,
+        escrow_seeds,
+        &crate::id(),
+        true,
+        INVALID_ESCROW_PDA,
+    )?;
+    load_owned_pda(escrow_account, &system_program::id(), INVALID_ESCROW_OWNER)?;
+
+    // deduce necessary accounts for CPI
+    let (accounts_meta, handler_accounts): (Vec<AccountMeta>, Vec<AccountInfo>) =
+        [escrow_authority_account, escrow_account]
+            .into_iter()
+            .chain(accounts_iter)
+            .filter(|account| account.key != validator.key)
+            .map(|account| {
+                (
+                    // We enable only escrow to be a signer
+                    AccountMeta {
+                        pubkey: *account.key,
+                        is_writable: account.is_writable,
+                        is_signer: account.key == escrow_account.key,
+                    },
+                    account.clone(),
+                )
+            })
+            .collect();
+    msg!(
+        "Calling, accounts_meta.len: {}, handler_account.len: {}",
+        accounts_meta.len(),
+        handler_accounts.len()
+    );
+
+    let data = [EXTERNAL_CALL_HANDLER_DISCRIMINATOR.to_vec(), data.to_vec()].concat();
+    let handler_instruction = Instruction {
+        program_id: *destination_program.key,
+        data,
+        accounts: accounts_meta,
+    };
+    let bump_slice = &[escrow_bump];
+    let escrow_signer_seeds = [escrow_seeds, &[bump_slice]].concat();
+
+    invoke_signed(
+        &handler_instruction,
+        &handler_accounts,
+        &[&escrow_signer_seeds],
+    )
+    .inspect_err(|err| {
```

Comment:
invoke_signed never returns an error, if the CPI fails, it would immediately abort and revert the transaction. To confirm that this is not critical, as it won't work as expected